File: aoc-py/solutions/day4.py
"""
Day 4: Scratchcards

https://adventofcode.com/2023/day/4
"""
__all__ = ('Day4',)

from ..solution import Solution

class Day4(Solution):
    NAME = 'Scratchcards'

    def _get_winning_amt(self, card: str) -> int:
        _, nums = card.split(':', maxsplit=1)
        winning, mine = nums.split('|', maxsplit=1)
        winning, mine = set(winning.split()), set(mine.split())

        return len(winning.intersection(mine))

    def part_one(self, inp: str) -> int:
        total = 0
        for card in inp.splitlines():
            amt_win = self._get_winning_amt(card)
            total += 2 ** (amt_win - 1) if amt_win > 0 else 0
        return total

    def part_two(self, inp: str) -> int:
        copies = [1] * len(lines := inp.splitlines())

        for i, (card, n_copies) in enumerate(zip(lines, copies)):
            win_amt = self._get_winning_amt(card)

            for card in range(i + 1, win_amt + i + 1):
                try:
                    copies[card] += n_copies
                except IndexError:
                    pass
        return sum(copies)

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 26914
        assert p2 == 13080971

File: src/bin/day4.rs
//! Day 4: Scratchcards
//!
//! <https://adventofcode.com/2023/day/4>
use std::{
    collections::HashSet,
    fmt::Display,
};
use aoc_2023::Solution;

pub struct Day4;

impl Day4 {
    /// returns the amount of the numbers that are winning numbers on a card
    fn get_winning_amt<T>(card: T) -> usize
    where
        T: AsRef<str>,
    {
        let string = card.as_ref();
        let (_, nums) = string
            .split_once(':')
            .unwrap();
        let (winning, mine) = nums
            .split_once('|')
            .unwrap();

        winning
            .split_whitespace()
            .collect::<HashSet<&str>>()
            .intersection(
                &mine
                    .split_whitespace()
                    .collect::<HashSet<&str>>()
            )
            .count()
    }

    /// # Panics
    ///
    /// If the amount of winning numbers exceeds [`u32::MAX`]
    pub fn part_one<T: Display>(&self, inp: T) -> u32 {
        inp
            .to_string()
            .lines()
            .map(|card| {
                let amt_win = Self::get_winning_amt(card);
                if amt_win > 0 {
                    (2_u32).pow(
                        u32::try_from(amt_win)
                            .unwrap() - 1
                    )
                } else {
                    0
                }
            })
            .sum()
    }

    pub fn part_two<T: Display>(&self, inp: T) -> u32 {
        let inp = inp.to_string();
        let mut copies = vec![1; inp.lines().count()];

        for (i, card) in inp
            .lines()
            .enumerate()
        {
            let win_amt = Self::get_winning_amt(card);

            for card in i + 1..=win_amt + i {
                let n_copies = *copies.get(i)
                    .unwrap_or(&0);
                if let Some(copy) = copies.get_mut(card) {
                    *copy += n_copies;
                }
            }
        }
        copies
            .into_iter()
            .sum()
    }
}

impl Solution for Day4 {
    const NAME: &'static str = "Scratchcards";

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 26914);
        assert_eq!(p2, 13_080_971);
    }
}

fn main() {
    aoc_2023::run_day(4, &Day4);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}